extendr-api/src/na.rs

```diff
@@ -0,0 +1,76 @@
+use crate::Bool;
+use libR_sys::{R_IsNA, R_NaReal};
+/// Return true if this primitive is NA.
+pub trait CanBeNA {
+    fn is_na(&self) -> bool;
+    fn na() -> Self;
+}
+
+/// ```
+/// use extendr_api::prelude::*;
+/// test! {
+///     assert!(<f64>::na().is_na());
+/// }
+/// ```
+impl CanBeNA for f64 {
+    fn is_na(&self) -> bool {
+        unsafe { R_IsNA(*self) != 0 }
+    }
+
+    fn na() -> f64 {
+        unsafe { R_NaReal }
+    }
+}
+
+/// ```
+/// use extendr_api::prelude::*;
+/// test! {
+///     assert!(<i32>::na().is_na());
+/// }
+/// ```
+impl CanBeNA for i32 {
+    fn is_na(&self) -> bool {
+        *self == i32::na()
+    }
+
+    fn na() -> i32 {
+        i32::MIN
+    }
+}
+
+/// ```
+/// use extendr_api::prelude::*;
+/// test! {
+///     assert!(<Bool>::na().is_na());
+/// }
+/// ```
+impl CanBeNA for Bool {
+    fn is_na(&self) -> bool {
+        self.0 == Bool::na().0
+    }
+
+    fn na() -> Bool {
+        Bool::from(i32::MIN)
+    }
+}
+
+/// Special "NA" string that represents null strings.
+/// ```
+/// use extendr_api::prelude::*;
+/// test! {
+///     assert_ne!(<&str>::na().as_ptr(), "NA".as_ptr());
+///     assert_eq!(<&str>::na(), "NA");
+///     assert_eq!("NA".is_na(), false);
+///     assert_eq!(<&str>::na().is_na(), true);
+/// }
+/// ```
+impl CanBeNA for &str {
+    /// Check for NA in a string by address.
+    fn is_na(&self) -> bool {
+        self.as_ptr() == <&str>::na().as_ptr()
```

Comment:
I know this was copied over from the code that was there earlier, but I don't understand how this works. We're comparing whether the pointer is equal to a special `NA` string we have made. But how would this work for `NA`s generated by R?

Reply:
Oh, I get it, this is for the Rust type `&str`, not for R character vectors.

Reply:
The concept here as far as I understand is that we create a special string. Original `na_str()` returned `&'static str`, so I assume that this special string is stored as a binary resource and our method always returns a slice from the same block of memory. As a result, by value this slice is identical to `"NA"`, but their in memory locations are different. So if we compare using ptr, only `"NA"` that was created using our method will have correct ptr and will test positive for `NA`, while any other string won't. 

This is a bit of a hack though, but my goal was to update trait, not to rethink implementation, which will likely change as soon as we get scalar and vector types for strings. Same applies for `Bool` used here.